crates/kitsune_p2p/fetch/src/respond.rs

```diff
@@ -0,0 +1,159 @@
+use kitsune_p2p_types::{KOpData, KSpace};
+use std::sync::Arc;
+
+/// Drop this when response sending is complete.
+pub struct FetchResponseGuard(tokio::sync::oneshot::Sender<()>);
+
+/// Customization by code making use of the FetchResponseQueue.
+pub trait FetchResponseConfig: 'static + Send + Sync {
+    /// Data that is forwarded.
+    type User: 'static + Send;
+
+    /// Byte count allowed to be outstanding.
+    /// Any ops requested to be enqueued over this amount
+    /// will be dropped without responding.
+    fn byte_limit(&self) -> u32 {
+        64 * 1024 * 1024
+    }
+
+    /// Number of concurrent sends to allow.
+    fn concurrent_send_limit(&self) -> u32 {
+        1
+    }
+
+    /// Send this fetch response.
+    fn respond(
+        &self,
+        space: KSpace,
+        user: Self::User,
+        completion_guard: FetchResponseGuard,
+        op: KOpData,
+    );
+}
+
+/// Manage responding to requests for data.
+pub struct FetchResponseQueue<C: FetchResponseConfig> {
+    byte_limit: Arc<tokio::sync::Semaphore>,
+    concurrent_send_limit: Arc<tokio::sync::Semaphore>,
+    config: Arc<C>,
+}
+
+impl<C: FetchResponseConfig> FetchResponseQueue<C> {
+    /// Construct a new response queue.
+    pub fn new(config: C) -> Self {
+        let byte_limit = Arc::new(tokio::sync::Semaphore::new(config.byte_limit() as usize));
+        let concurrent_send_limit = Arc::new(tokio::sync::Semaphore::new(
+            config.concurrent_send_limit() as usize,
+        ));
+        let config = Arc::new(config);
+        Self {
+            byte_limit,
+            concurrent_send_limit,
+            config,
+        }
+    }
+
+    /// Enqueue an op to be sent to a remote.
+    pub fn enqueue_op(&self, space: KSpace, user: C::User, op: KOpData) {
+        let len = op.size();
+
+        if len > u32::MAX as usize {
+            tracing::warn!("op size > u32::MAX");
```

Comment:
This would be really unusual

```suggestion
            tracing::error!("op size > u32::MAX");
```